### DISCUSSION TITLE
Public API for SymbolsModel

### DISCUSSION THREAD

**Discussion by ghentschke** [eclipse/lsp4e#630] [Ideas]
Hi,
I'am currently working on a LS based Project Explorer View. The children of a Translation Unit / Document shall be fetched from the LS. 
I reused the `org.eclipse.lsp4e.outline.SymbolsModel` for it. 

Are there any  plans to provide a stable API for the SymbolsModel? 
I think there a several possible application which uses a TreeView for a Translation Unit/Document (Outline,  Project Explorer, ...) and it makes sense to reuse the data model.

**Comment by mickaelistria** [comment]
If it's for the Project Explorer, then it means it's using the common navigator framework (CNF) and LSP4E already provides extensions for this framework (that would only need to be enabled for the Project Explorer).
So I think we can instead of opening API, just try to make things work well for all CNF-based viewers in general. Maybe we just miss some adapter to get the right document from the parent?

**Reply by aubade** [reply]
Sorry to necropost and hijack the thread, but is there any documentation on how to actually enable the CNF extensions for the project explorer? (or is this something that has to happen on the LSP4e side?)

I'm doing some research about possibly reviving the old D Development Tools project and possibly migrating it to use LSP4e. And making the source files expandable in the project explorer is one of the most important new features to me.

**Reply by mickaelistria** [reply]
Ideally such contribution to common navigator would be implemented in LSP4E, so it would benefit all files that have a language server bound. Documentation about CNF is at https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2Fcnf.htm . As mentioned in the other post, playing with trigger points and adapters may allow symbols to be visible from project explorer.

However, I personally prefer having symbols in the outline, out of the Project Explorer: I find the project explorer easily gets too crowded and less usable when it gets under a grain finer than files. So IMO, if you're working on adopting a new language server, I don't think it's worth making the topic of symbols in project explorer a high priority for a few iterations.

**Reply by aubade** [reply]
Thanks! It'll be a few iterations to get there anyway. The original DDT project was abandoned in 2017, so I have to first make it work with modern Eclipse and adapt to LSP4e instead of building off CDT directly.

**Comment by ghentschke** [comment]
> If it's for the Project Explorer, then it means it's using the common navigator framework (CNF) and LSP4E already provides extensions for this framework (that would only need to be enabled for the Project Explorer).

Do you mean the `org.eclipse.lsp4e.outline.content` navigatorContent extension ? 
I think this is not applicable to a language specific project TreeView in the Project Explorers view, since it covers only elements which children consists of symbols (e.g. a translation unit document). The existing C/C++ navigatorContent covers a complex C/C++ language specific data model which should be expanded by a new symbols provider for a given Translation Unit as parent.  Here comes the language server into account. It could extend existing `navigatorContent` views by using `triggerPoints` where the element can be defined for which the `navigatorContent` should be used to determine the children for:
![image](https://user-images.githubusercontent.com/123444711/236827291-cc727dd2-29f0-4485-aa06-7d3486fb480f.png)

**Reply by mickaelistria** [reply]
I there some common type a translation unit can adapt to (eg `IDocument`, `IResource`, `URI`...) ? If so, we can make the outline extension check for adaptable trigger point.
